head/src/main/kotlin/io/qalipsis/core/head/lock/RedisLockProviderImpl.kt

```diff
@@ -0,0 +1,57 @@
+/*
+ * QALIPSIS
+ * Copyright (C) 2022 AERIS IT Solutions GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package io.qalipsis.core.head.lock
+
+import io.aerisconsulting.catadioptre.KTestable
+import io.lettuce.core.RedisClient
+import io.micronaut.context.annotation.Requirements
+import io.micronaut.context.annotation.Requires
+import io.qalipsis.api.context.CampaignKey
+import io.qalipsis.core.configuration.ExecutionEnvironments
+import jakarta.inject.Singleton
+import java.time.Duration
+import ru.sokomishalov.lokk.provider.RedisLettuceLokkProvider
+import ru.sokomishalov.lokk.provider.withLokk
+
+/**
+ * Custom redis implementation of the lockProvider.
+ *
+ * @author Francisca Eze
+ */
+@Singleton
+@Requirements(
+    Requires(env = [ExecutionEnvironments.HEAD]),
+    Requires(notEnv = [ExecutionEnvironments.SINGLE_HEAD])
+)
+internal class RedisLockProviderImpl(private val redisClient: RedisClient) : LockProvider {
+
+    @KTestable
+    private val redisLettuceLockProvider = RedisLettuceLokkProvider(client = redisClient)
+
+
+    override suspend fun withLock(campaignKey: CampaignKey, block: suspend () -> Unit) {
+        redisLettuceLockProvider.withLokk(
+            name = campaignKey,
+            atLeastFor = Duration.ofMillis(1),
+            atMostFor = Duration.ofMillis(20),
```

Comment:
does that mean that en error is returned after 20 ms if the lock could not be acquired? 2 seconds seems safer.

Reply:
Yea I would like to have a conversation with you about these values